### PR TITLE
Timeout

### PR DESCRIPTION
**Pull request by andreikop**
This PR adds ability to set request timeout for all API functions.
There are no timeout in curl by default.
Without timeout our daemon often hangs forever when there are some problems with the S3 service.

```
$ build/bin/s3 -r 0 list                                    
                         Bucket                                 Created       
--------------------------------------------------------  --------------------
andreikop-test                                            2017-02-03T10:50:04Z
$ build/bin/s3 -r 0 -t 1000 list
                         Bucket                                 Created       
--------------------------------------------------------  --------------------
andreikop-test                                            2017-02-03T10:50:04Z
$ build/bin/s3 -r 0 -t 10 list  

ERROR: ErrorRequestTimeout
```

### CODE REVIEW

File: inc/request.h [hunk 1]
```diff
@@ -102,6 +102,9 @@ typedef struct RequestParams
 
     // Data passed to the callbacks
     void *callbackData;
+
+    // Request timeout. 0 if not set
```
Comment (bji) [inline inc/request.h]: I think this "0 if not set" comment is inaccurate and confusing?  Or maybe I just don't understand?  What about "If 0, no timeout will be enforced"?

File: src/acl.c [hunk 2]
```diff
@@ -270,7 +272,7 @@ static int setAclDataCallback(int bufferSize, char *buffer, void *callbackData)
 }
 
 
-static void setAclCompleteCallback(S3Status requestStatus,
```
Comment (bji) [inline src/acl.c]: I really have to figure out why my emacs configuration doesn't tidy up extraneous spaces automatically ...

**Comment by bji** [comment]
You'll have to forgive me but it's been so long since I worked on this library ... there is a multiplexed I/O mode isn't there?  If I recall correctly you can either make your requests synchronously or asynchronously via the request context ... right?  So I'm wondering, are the timeouts enforced for both synchronous and asynchronous modes (curl_easy and curl_multi)?  Did you test that?

**Comment by andreikop** [comment]
Yes, I've tested curl_multi. It works the same way.

File: inc/request.h (hunk 1 above)
Comment (andreikop) [inline inc/request.h]: I've updated the comment

**Comment by bji** [comment]
Thank you for your great work!

**Pull request merged**